### PR TITLE
Refactor multi tensor indexing

### LINKED ISSUE #235: Refactor advanced indexing in `MultiEmbeddingTensor` and `MultiNestedTensor`
Do you think `_slice`, `narrow`, `select`, `index_select` can be shared with `MultiNestedTensor` (in other words, we put them in the parent `_MultiTensor` class)? Here we only need to implement `_row_index_select`, `_col_index_select` and so on.

_Originally posted by @weihua916 in https://github.com/pyg-team/pytorch-frame/pull/217#discussion_r1397971207_

### PR DESCRIPTION
**Pull request by akihironitta**
Addresses https://github.com/pyg-team/pytorch-frame/pull/217#discussion_r1397971207. Closes #235.

### CODE REVIEW

File: torch_frame/data/multi_embedding_tensor.py [hunk 1]
```diff
@@ -96,186 +96,37 @@ def from_tensor_list(
...
-        if start == 0 and start + length >= num_data:
-            return self
-        elif length <= 0:
-            return MultiEmbeddingTensor(
-                num_rows=0 if dim == 0 else self.num_rows,
-                num_cols=0 if dim == 1 else self.num_cols,
-                values=torch.tensor([], device=self.device),
-                offset=torch.tensor([0], device=self.device)
-                if dim == 1 else self.offset,
-            )
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/data/multi_embedding_tensor.py:212]: Hmm, that's a great point. I think having the two distinct behaviors is the right thing to do here.

File: torch_frame/data/multi_tensor.py [hunk 2]
```diff
@@ -189,6 +189,109 @@ def allclose(
...
+        if start == 0 and start + length >= num_data:
+            return self
+        elif length <= 0:
+            return self.__class__(
+                num_rows=0 if dim == 0 else self.num_rows,
+                num_cols=0 if dim == 1 else self.num_cols,
+                values=torch.tensor([], device=self.device),
+                offset=torch.tensor([0], device=self.device)
+                if dim == 1 else self.offset,
+            )
```
Comment (weihua916) [inline torch_frame/data/multi_tensor.py:268]: I think we will need a separate functionalities for `MultiNestedTensor` versus `MultiEmbeddedTensor`.
How about we add
```python
def empty_tensor(self, dim: int) -> _MultiTensor:
   raise NotImplementedError
    ...
```
and return `self.empty_tensor(dim=dim)` here.

We then implement `empty_tensor` function for `MultiNestedTensor` and `MultiEmbeddingTensor` separately.

Comment (akihironitta) [reply torch_frame/data/multi_tensor.py]: Done! PTAL when you get a chance :)

File: torch_frame/data/multi_tensor.py [hunk 3]
```diff
@@ -189,6 +189,137 @@ def allclose(
...
+            return self._col_narrow(start, length)
+
+    def _row_narrow(self, start: int, length: int) -> "_MultiTensor":
+        raise NotImplementedError
+
+    def _col_narrow(self, start: int, length: int) -> "_MultiTensor":
+        raise NotImplementedError
+
+    def empty(self, dim: int) -> "_MultiTensor":
+        raise NotImplementedError
```
Comment (weihua916) [inline torch_frame/data/multi_tensor.py:288]: Shall we make it `_empty`, since it's more of a helper function and not expected to be called externally?

File: torch_frame/data/multi_nested_tensor.py [hunk 4]
```diff
@@ -400,6 +290,24 @@ def to_dense(self, fill_value: Union[int, float]) -> Tensor:
         dense[row, col, arange] = self.values
         return dense
 
+    def empty(self, dim: int) -> "MultiNestedTensor":
+        """Creates an empty :class:`MultiEmbeddingTensor`.
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:294]: nit: Add r"""

**Review by weihua916 — APPROVED** [review]
Thanks! LGTM after addressing the final nits.
